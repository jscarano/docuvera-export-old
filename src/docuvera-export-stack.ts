import {
  CorsHttpMethod,
  HttpApi,
  HttpMethod,
} from '@aws-cdk/aws-apigatewayv2-alpha';
import { AttributeType, BillingMode, Table } from 'aws-cdk-lib/aws-dynamodb';
import { Architecture } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Construct } from 'constructs';
import { HttpLambdaIntegration } from '@aws-cdk/aws-apigatewayv2-integrations-alpha';
import {
  RemovalPolicy,
  Stack,
  StackProps,
} from 'aws-cdk-lib';
// import {
//   CfnOutput,
//   DockerImage,
//   RemovalPolicy,
//   Stack,
//   StackProps,
// } from 'aws-cdk-lib';
// import {
//   Distribution,
//   OriginAccessIdentity,
//   ViewerProtocolPolicy,
// } from 'aws-cdk-lib/aws-cloudfront';
// import { S3Origin } from 'aws-cdk-lib/aws-cloudfront-origins';
// import { PolicyStatement } from 'aws-cdk-lib/aws-iam';
// import { RetentionDays } from 'aws-cdk-lib/aws-logs';
// import { BlockPublicAccess, Bucket } from 'aws-cdk-lib/aws-s3';
// import { BucketDeployment, Source } from 'aws-cdk-lib/aws-s3-deployment';
// import {
//   AwsCustomResource,
//   AwsCustomResourcePolicy,
//   PhysicalResourceId,
// } from 'aws-cdk-lib/custom-resources';
// import { execSync, ExecSyncOptions } from 'child_process';

// import { copySync } from 'fs-extra';
// import { join } from 'path';

export class DocuveraExportStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // Demo-quality props. For production, you want a different removalPolicy and possibly a different billingMode.
    const table = new Table(this, 'NotesTable', {
      billingMode: BillingMode.PAY_PER_REQUEST,
      partitionKey: { name: 'pk', type: AttributeType.STRING },
      removalPolicy: RemovalPolicy.DESTROY,
      sortKey: { name: 'sk', type: AttributeType.STRING },
      tableName: 'NotesTable',
    });

    // Functions could have memory tuned to save $$, but should be pretty cheap in any case.
    const readFunction = new NodejsFunction(this, 'ReadNotesFn', {
      architecture: Architecture.ARM_64,
      entry: `${__dirname}/lambdas/readFunction.ts`
    });

    const writeFunction = new NodejsFunction(this, 'WriteNoteFn', {
      architecture: Architecture.ARM_64,
      entry: `${__dirname}/lambdas/writeFunction.ts`
    });


    table.grantReadData(readFunction);
    table.grantWriteData(writeFunction);


    const convertFunction = new NodejsFunction(this, 'ConvertFn', {
      architecture: Architecture.ARM_64,
      entry: `${__dirname}/lambdas/convertFunction.ts`
    });

    // API could be improved with authorization and models to validate payloads.
    // In production, you will want access logging.
    const api = new HttpApi(this, 'NotesApi', {
      corsPreflight: {
        allowHeaders: ['Content-Type'],
        allowMethods: [CorsHttpMethod.GET, CorsHttpMethod.POST],
        allowOrigins: ['*'],
      },
    });

    // Creates the Cfn AWS::ApiGatewayV2::Integration resources
    const readIntegration = new HttpLambdaIntegration(
      'ReadIntegration',
      readFunction
    );
    const writeIntegration = new HttpLambdaIntegration(
      'WriteIntegration',
      writeFunction
    );

    // Creates the Cfn AWS::ApiGatewayV2::Route resources, assigning a path to an integration
    api.addRoutes({
      integration: readIntegration,
      methods: [HttpMethod.GET],
      path: '/notes',
    });
    api.addRoutes({
      integration: writeIntegration,
      methods: [HttpMethod.POST],
      path: '/notes',
    });

    // // Storage for assets only. NOT an S3 website.
    // const websiteBucket = new Bucket(this, 'WebsiteBucket', {
    //   autoDeleteObjects: true,
    //   blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
    //   removalPolicy: RemovalPolicy.DESTROY,
    // });

    // // Grant read access to the distribution.
    // const originAccessIdentity = new OriginAccessIdentity(
    //   this,
    //   'OriginAccessIdentity'
    // );
    // websiteBucket.grantRead(originAccessIdentity);

    // Cloudfront distribution with SPA config and https upgrade.
    // const distribution = new Distribution(this, 'Distribution', {
    //   defaultBehavior: {
    //     origin: new S3Origin(websiteBucket, { originAccessIdentity }),
    //     viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
    //   },
    //   defaultRootObject: 'index.html',
    //   errorResponses: [
    //     {
    //       httpStatus: 404,
    //       responseHttpStatus: 200,
    //       responsePagePath: '/index.html',
    //     },
    //   ],
    // });

    // const execOptions: ExecSyncOptions = {
    //   stdio: ['ignore', process.stderr, 'inherit'],
    // };

    // Run vite build to transpile React application, then copy to cdk.out.
    // Docker build can't be omitted, even though we don't use it.
    // const bundle = Source.asset(join(__dirname, 'web'), {
    //   bundling: {
    //     command: [
    //       'sh',
    //       '-c',
    //       'echo "Docker build not supported. Please install esbuild."',
    //     ],
    //     image: DockerImage.fromRegistry('alpine'),
    //     local: {
    //       tryBundle(outputDir: string) {
    //         try {
    //           execSync('esbuild --version', execOptions);
    //         } catch {
    //           return false;
    //         }
    //         execSync('npx vite build', execOptions);
    //         copySync(join(__dirname, '../dist'), outputDir, {
    //           ...execOptions,
    //           recursive: true,
    //         });
    //         return true;
    //       },
    //     },
    //   },
    // });

    // Need to set prune to false or the config.json file will be pruned.
    // If deployments are frequent, should look into a way to clean up old files.
    // new BucketDeployment(this, 'DeployWebsite', {
    //   destinationBucket: websiteBucket,
    //   distribution,
    //   logRetention: RetentionDays.ONE_DAY,
    //   prune: false,
    //   sources: [bundle],
    // });

    // Generate a config.json file and place in S3 so the web app can grab the API URL.
    // new AwsCustomResource(this, 'ApiUrlResource', {
    //   logRetention: RetentionDays.ONE_DAY,
    //   onUpdate: {
    //     action: 'putObject',
    //     parameters: {
    //       Body: Stack.of(this).toJsonString({
    //         [this.stackName]: { HttpApiUrl: api.apiEndpoint },
    //       }),
    //       Bucket: websiteBucket.bucketName,
    //       CacheControl: 'max-age=0, no-cache, no-store, must-revalidate',
    //       ContentType: 'application/json',
    //       Key: 'config.json',
    //     },
    //     physicalResourceId: PhysicalResourceId.of('config'),
    //     service: 'S3',
    //   },
    //   policy: AwsCustomResourcePolicy.fromStatements([
    //     new PolicyStatement({
    //       actions: ['s3:PutObject'],
    //       resources: [websiteBucket.arnForObjects('config.json')],
    //     }),
    //   ]),
    // });

    // new CfnOutput(this, 'HttpApiUrl', { value: api.apiEndpoint });

    // new CfnOutput(this, 'DistributionDomain', {
    //   value: distribution.distributionDomainName,
    // });
  }
}
