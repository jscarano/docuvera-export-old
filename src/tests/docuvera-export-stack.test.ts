import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

import { DocuveraExportStack } from '../docuvera-export-stack';

describe('Stack test', () => {
  test('snapshot', () => {
    const app = new App();
    const stack = new DocuveraExportStack(app, 'TestStack');
    const cfn = Template.fromStack(stack).toJSON();
    const resources = cfn.Resources;
    const matchObject: {
      Parameters: Record<string, unknown>;
      Resources: Record<string, unknown>;
    } = {
      Parameters: expect.any(Object),
      Resources: {},
    };

    /**
     * Exact matches on the code artifacts product lots of churn.
     * Doing this produces more stable tests as the expense of some accuracy.
     * Prefer this over continually updating snapshots.
     */
    Object.keys(resources).forEach((res) => {
      switch (resources[res].Type) {
        case 'AWS::Lambda::Function':
          matchObject.Resources[res] = {
            Properties: { Code: expect.any(Object) },
          };
          break;
        case 'Custom::CDKBucketDeployment':
          matchObject.Resources[res] = {
            Properties: {
              SourceBucketNames: expect.any(Array),
              SourceObjectKeys: expect.any(Object),
            },
          };
          break;
        default:
          break;
      }
    });
    expect(cfn).toMatchSnapshot(matchObject);
    // Optionally add additional assertions to count resources, etc.
  });
});
