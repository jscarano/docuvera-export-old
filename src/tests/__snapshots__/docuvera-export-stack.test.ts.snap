// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack test snapshot 1`] = `
Object {
  "Parameters": Any<Object>,
  "Resources": Object {
    "ConvertFnCAE452AA": Object {
      "DependsOn": Array [
        "ConvertFnServiceRoleA324E8EB",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "ConvertFnServiceRoleA324E8EB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ConvertFnServiceRoleA324E8EB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "NotesApi2FFCBC92": Object {
      "Properties": Object {
        "CorsConfiguration": Object {
          "AllowHeaders": Array [
            "Content-Type",
          ],
          "AllowMethods": Array [
            "GET",
            "POST",
          ],
          "AllowOrigins": Array [
            "*",
          ],
        },
        "Name": "NotesApi",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "NotesApiDefaultStage018487EA": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "NotesApi2FFCBC92",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "NotesApiGETnotes1ADF0919": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "NotesApi2FFCBC92",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "GET /notes",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "NotesApiGETnotesReadIntegration044C0BBC",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "NotesApiGETnotesReadIntegration044C0BBC": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "NotesApi2FFCBC92",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::GetAtt": Array [
            "ReadNotesFn9B772BEA",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "NotesApiGETnotesReadIntegrationPermission44DDAC24": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "ReadNotesFn9B772BEA",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "NotesApi2FFCBC92",
              },
              "/*/*/notes",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "NotesApiPOSTnotes6D069734": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "NotesApi2FFCBC92",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "POST /notes",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "NotesApiPOSTnotesWriteIntegration94DB7B6D",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "NotesApiPOSTnotesWriteIntegration94DB7B6D": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "NotesApi2FFCBC92",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::GetAtt": Array [
            "WriteNoteFn75A93DA3",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "NotesApiPOSTnotesWriteIntegrationPermissionCE0F7A77": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "WriteNoteFn75A93DA3",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "NotesApi2FFCBC92",
              },
              "/*/*/notes",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "NotesTableD0D0D2F1": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "pk",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "sk",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "pk",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "sk",
            "KeyType": "RANGE",
          },
        ],
        "TableName": "NotesTable",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "ReadNotesFn9B772BEA": Object {
      "DependsOn": Array [
        "ReadNotesFnServiceRoleDefaultPolicyF0CAA4ED",
        "ReadNotesFnServiceRole4182CAE7",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "ReadNotesFnServiceRole4182CAE7",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ReadNotesFnServiceRole4182CAE7": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ReadNotesFnServiceRoleDefaultPolicyF0CAA4ED": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "NotesTableD0D0D2F1",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ReadNotesFnServiceRoleDefaultPolicyF0CAA4ED",
        "Roles": Array [
          Object {
            "Ref": "ReadNotesFnServiceRole4182CAE7",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WriteNoteFn75A93DA3": Object {
      "DependsOn": Array [
        "WriteNoteFnServiceRoleDefaultPolicyE6121555",
        "WriteNoteFnServiceRoleEC618145",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "WriteNoteFnServiceRoleEC618145",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "WriteNoteFnServiceRoleDefaultPolicyE6121555": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "NotesTableD0D0D2F1",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WriteNoteFnServiceRoleDefaultPolicyE6121555",
        "Roles": Array [
          Object {
            "Ref": "WriteNoteFnServiceRoleEC618145",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WriteNoteFnServiceRoleEC618145": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
